brave/src/main/java/brave/propagation/B3SingleFormat.java

```diff
@@ -73,49 +97,83 @@ static String writeB3SingleFormat(TraceContext context) {
       return null;
     }
 
+    int pos = 0;
     long traceIdHigh, traceId;
-    boolean traceId128 = b3.charAt(32) == '-';
-    if (traceId128) {
-      traceIdHigh = lenientLowerHexToUnsignedLong(b3, 0, 16);
-      traceId = lenientLowerHexToUnsignedLong(b3, 16, 32);
+    if (b3.charAt(32) == '-') {
+      traceIdHigh = tryParse16HexCharacters(b3, pos, length);
+      pos += 16; // upper 64 bits of the trace ID
+      traceId = tryParse16HexCharacters(b3, pos, length);
     } else {
       traceIdHigh = 0L;
-      traceId = lenientLowerHexToUnsignedLong(b3, 0, 16);
+      traceId = tryParse16HexCharacters(b3, pos, length);
     }
+    pos += 16; // traceId
+    if (!checkHyphen(b3, pos)) return null;
+    pos++; // consume the hyphen
 
     if (traceIdHigh == 0L && traceId == 0L) {
       logger.fine("Invalid input: expected a 16 or 32 lower hex trace ID at offset 0");
       return null;
     }
 
-    int pos = traceId128 ? 33 : 17; // traceid-
-    long spanId = lenientLowerHexToUnsignedLong(b3, pos, pos + 16);
+    long spanId = tryParse16HexCharacters(b3, pos, length);
     if (spanId == 0L) {
       logger.log(FINE, "Invalid input: expected a 16 lower hex span ID at offset {0}", pos);
       return null;
     }
-    pos += 17; // spanid-
+    pos += 16; // spanid
 
     int flags = 0;
-    if (length == pos + 1 || (length > pos + 2 && b3.charAt(pos + 1) == '-')) {
-      flags = parseSampledFlag(b3, pos++);
-      if (flags == 0) return null;
-      pos++; // skip the dash
-    }
-
     long parentId = 0L;
-    if (length >= pos + 17) {
-      parentId = lenientLowerHexToUnsignedLong(b3, pos, pos + 16);
-      if (parentId == 0L) {
-        logger.log(FINE, "Invalid input: expected a 16 lower hex parent ID at offset {0}", pos);
+    if (length > pos) {
+      // If we are at this point, we have more than just traceId-spanId.
+      // If the sampling field is present, we'll have a delimiter 2 characters from now. Ex "-1"
+      // If it is absent, but a parent ID is (which is strange), we'll have at least 17 characters.
+      // Therefore, if we have less than two characters, the input is truncated.
+      if (length == pos + 1) {
+        logger.fine("Invalid input: truncated");
```

Comment:
another idea I have is to make a base type which accepts a "parse error" operation. by default, this can go to JUL, but it could also go into a place where it can be a span tag.. for example propagation fail to parse reason.